config,lightning: Implements server mode

kennytm opened the pull request:
<!--
Thank you for contributing to TiDB-Lightning! Please read the [CONTRIBUTING](https://github.com/pingcap/tidb-lightning/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->

As a prerequisite of the "web interface", we'd like to have Lightning keep running even after import is finished. 

The behavior overlaps with a feature request which turns Lightning into a server, which we could submit tasks to it (similar to DM-master). The conversion is straightforward, so why not 🙃.

### What is changed and how it works?

1. Split config into two classes: "Global config" and "Task config" (both  are subset of `tidb-lightning.toml`, unlike DM task settings). 
    * Also refactored the existing argument parsing code to cut down the code duplication.

2. Added an HTTP endpoint which allows user to upload a Task config toml. 

3. Changed the public `Run()` method to loop for those uploaded tasks in server mode. Essentially like this:

    ```go
    func Run() {
        if !ServerMode {
            run(globalConfig)
        } else {
            for taskConfig := range uploadedTaskConfigs {
                run(taskConfig)
            }
        }
    }
    ```

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test
 - Integration test

Side effects

 - Increased code complexity

Related changes

 - Need to update the documentation 
 - Need to be included in the release note

kennytm commented:
/run-all-tests

kennytm commented:
/run-all-tests

kennytm commented:
/run-all-tests

GregoryIan commented:
@july2993 @lonng PTAL

lonng commented on lightning/config/global.go:
```diff
@@ -0,0 +1,183 @@
+	ServerMode bool   `toml:"server-mode" json:"server-mode"`
+
+	// The legacy alias for setting "status-addr"
+	PProfPort int `toml:"pprof-port" json:"-"`
```
I think it's better to add some comments about the reason for ignoring the `PProfPort` field in the `JSON` configuration. Otherwise, we should not ignore it.

lonng commented on cmd/tidb-lightning/main.go:
```diff
@@ -38,16 +36,7 @@ func setGlobalVars() {
 func main() {
 	setGlobalVars()
```
Should we remove this function? The `PreparedPlanCache` feature has some extra cost but seems we never rely on this feature.

lonng commented on lightning/lightning.go:
```diff
@@ -32,67 +34,105 @@ import (
+var taskCfgRecorderKey struct{}
+
+func (l *Lightning) run(taskCfg *config.Config) error {
+	failpoint.Inject("SkipRunTask", func() {
```
We can change the failpoint clousure signature to remove the `failpoint.Return` call.
```go
	failpoint.Inject("SkipRunTask", func() error {
		if recorder, ok := l.ctx.Value(&taskCfgRecorderKey).(chan *config.Config); ok {
			recorder <- taskCfg
		}
		return nil
	})
```

lonng commented on lightning/lightning.go:
```diff
@@ -32,67 +34,105 @@ import (
-			log.L().Info("stopped HTTP server", log.ShortError(err))
-		}()
-	}
+	runtime.GOMAXPROCS(runtime.NumCPU())
```
It's not necessary to call `runtime.GOMAXPROCS` manually.

lonng commented on lightning/lightning.go:
```diff
@@ -105,6 +145,59 @@ func (l *Lightning) run() error {
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte("task queued"))
+	default:
+		http.Error(w, "task queue overflowed, please try again later", http.StatusServiceUnavailable)
```
The execution flow will reach here if the channel was closed, do we need to change the error message to adapt it. (Although we don't close the channel in the current implementation)

july2993 commented on lightning/lightning.go:
```diff
@@ -32,67 +34,105 @@ import (
+		select {
+		case <-l.ctx.Done():
+			return l.ctx.Err()
+		case task := <-l.taskCfgs:
```
seems nowhere will close l.taskCfgs now? and will not push nil(should not)

kennytm commented:
/run-all-tests

kennytm commented:
PTAL again @lonng @july2993 

Changed the queue data structure from a bare channel to a linked hash map to allow removing an item from the queue, and removed the artificial queue size limit. The output format of `/task` is changed to JSON.

GregoryIan commented:
@lonng PTAL again

GregoryIan commented:
it would be easier to integrate into TiDB, right?

lonng commented on lightning/config/configlist.go:
```diff
@@ -0,0 +1,117 @@
+
+	select {
+	case cfg := <-res:
+		return cfg, nil
```
We need to close the `channel` before returning.

kennytm replied:
Closing a channel is not necessary https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open

kennytm commented:
@GregoryIan as a plug-in embedding the whole app, probably just the same difficulty.

GregoryIan commented:
should we add task ID into some component (like checkpoint and metrics)?

kennytm commented:
Checkpoint no, metrics maybe, logs yes.

GregoryIan commented:
if we finish each job, I think we should clear checkpoint or initial new checkpoint with new task id

GregoryIan commented:
discuss with @kennytm, we would use `time` as task ID, and add task ID into checkpoint database/file, and log

GregoryIan commented:
LGTM

kennytm commented:
/run-all-tests

kennytm commented:
PTAL @GregoryIan
